File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.cpp
// https://leetcode.com/problems/letter-combinations-of-a-phone-number

class Solution {
public:
    vector<string> letterCombinations(string digits) {
        const vector<string>pad={"","","abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"};
        if(digits.empty())return{};
        vector<string>res={""};
        for(char digit:digits){
            vector<string>tmp;
            for(char letter:pad[digit-'0'])
                for(string s:res)
                    tmp.push_back(s+letter);
            res=tmp;
        }
        return res;
    }
};

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.go
func letterCombinations(digits string) []string {
	if len(digits) == 0 {
		return nil
	}
	pad := []string{"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"}
	res := []string{""}
	for _, digit := range digits {
		tmp := []string{}
		letters := pad[digit-'0']
		for _, letter := range letters {
			for _, s := range res {
				tmp = append(tmp, s+string(letter))
			}
		}
		res = tmp
	}
    return res
}